### PR TITLE
gh-468 Display data decoded

### PR DESCRIPTION
**Pull request by MouazAlzahabi**

### CODE REVIEW

File: Multisig/Data/Services/Safe Client Gateway Service/SCGModels.swift [hunk 1]
```diff
@@ -263,6 +263,21 @@ extension DataDecodedParameter {
 
 protocol DataDecodedParameterValue {}
 
+extension DataDecodedParameterValue {
+    var stringValue: String? {
+        return self as? String
```
Comment (DmitryBespalov) [inline Multisig/Data/Services/Safe Client Gateway Service/SCGModels.swift]: `return` can be skipped

File: Multisig/Data/Services/Safe Client Gateway Service/SCGModels.swift [hunk 2]
```diff
@@ -263,6 +263,21 @@ extension DataDecodedParameter {
...
+    }
+
+    var addressValue: Address? {
+        guard let stringValue = stringValue else { return nil }
+        return Address(stringValue)
+    }
+
+    var arrayValue: [DataDecodedParameterValue]? {
+        return self as? [DataDecodedParameterValue]
```
Comment (DmitryBespalov) [inline Multisig/Data/Services/Safe Client Gateway Service/SCGModels.swift]: `return` can be removed

File: Multisig/Logic/Models/Transaction/TransactionData.swift [hunk 3]
```diff
@@ -17,6 +17,17 @@ struct TransactionDataParameter: Hashable {
     let name: String
     let type: String
     let value: String?
+    let decodedData: DataDecodedParameterValue?
+
+    static func == (lhs: TransactionDataParameter, rhs: TransactionDataParameter) -> Bool {
+        return lhs.name == rhs.name && lhs.type == rhs.type && lhs.value == rhs.value
+    }
+
+    func hash(into hasher: inout Hasher) {
```
Comment (DmitryBespalov) [inline Multisig/Logic/Models/Transaction/TransactionData.swift:26]: 👍

File: Multisig/Logic/Models/Transaction/TransactionData.swift [hunk 4]
```diff
@@ -17,6 +17,17 @@ struct TransactionDataParameter: Hashable {
     let name: String
     let type: String
     let value: String?
+    let decodedData: DataDecodedParameterValue?
+
+    static func == (lhs: TransactionDataParameter, rhs: TransactionDataParameter) -> Bool {
+        return lhs.name == rhs.name && lhs.type == rhs.type && lhs.value == rhs.value
```
Comment (DmitryBespalov) [inline Multisig/Logic/Models/Transaction/TransactionData.swift]: `return` can be removed, this thing has another fancy  way of doing, whatever you prefer:

```
(lhs.name, lhs.type, lhs.value)  == (rhs.name, rhs.type, rhs.value)
```

File: Multisig/UI/Transactions/TransactionDetails/ParameterValueView.swift [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    @ViewBuilder
+    var valueView: some View {
+        if value?.addressValue != nil  {
+            AddressCell(address: value!.addressValue!.checksummed)
+        }
+        else if stringValue != nil  {
+            Text(stringValue!).body(.gnoDarkGrey)
+        } else {
+            Group {
+                ForEach((0..<value!.arrayValue!.count)) { index in
```
Comment (DmitryBespalov) [inline Multisig/UI/Transactions/TransactionDetails/ParameterValueView.swift:27]: we haven't tested this case  of  nested array values yet with  actual transaction

File: Multisig/UI/Transactions/TransactionDetails/TransactionDetailsView.swift [hunk 6]
```diff
@@ -59,6 +59,10 @@ struct TransactionDetailsView: Loadable {
         Group {
             TransactionHeaderView(transaction: transaction)
 
+            if transaction.dataDecoded != nil {
```
Comment (DmitryBespalov) [inline Multisig/UI/Transactions/TransactionDetails/TransactionDetailsView.swift]: it  should also check that the type of transaction is custom transaction

**Review by DmitryBespalov — CHANGES_REQUESTED** [review]

File: Multisig/UI/Transactions/TransactionDetails/ParameterValueView.swift [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+    var body: some View {
+        valueView.padding(.leading, CGFloat(nestingLevel) * 8)
+    }
+
+    @ViewBuilder
+    var valueView: some View {
+        if value?.addressValue != nil  {
+            AddressCell(address: value!.addressValue!.checksummed)
+        }
+        else if stringValue != nil  {
```
Comment (sche) [inline Multisig/UI/Transactions/TransactionDetails/ParameterValueView.swift]: tiny: could you put is on the same line, please?

File: Multisig/UI/Transactions/TransactionDetails/ParameterValueView.swift [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+                ForEach((0..<value!.arrayValue!.count)) { index in
+                    ParameterValueView(value: self.value!.arrayValue![index],
+                                       nestingLevel: self.nestingLevel + 1)
+                }
+            }
+        }
+    }
+
+    var stringValue: String? {
+        guard let value = value else { return "-"}
```
Comment (sche) [inline Multisig/UI/Transactions/TransactionDetails/ParameterValueView.swift]: space after "}"

**Review by sche — APPROVED** [review]

**Review by DmitryBespalov — APPROVED** [review]

**Review by sche — APPROVED** [review]

**Pull request merged**